rollup/rollupsyncservice/chunk.go

```diff
@@ -0,0 +1,155 @@
+	return numL1Messages
+}
+
+// Encode encodes the Chunk into RollupV2 Chunk Encoding.
+func (c *Chunk) Encode(totalL1MessagePoppedBefore uint64) ([]byte, error) {
+	numBlocks := len(c.Blocks)
+
+	if numBlocks > 255 {
+		return nil, errors.New("number of blocks exceeds 1 byte")
+	}
+	if numBlocks == 0 {
+		return nil, errors.New("number of blocks is 0")
+	}
+
+	var chunkBytes []byte
+	chunkBytes = append(chunkBytes, byte(numBlocks))
+
+	var l2TxDataBytes []byte
+
+	for _, block := range c.Blocks {
+		blockBytes, err := block.Encode(totalL1MessagePoppedBefore)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode block: %v", err)
+		}
+		totalL1MessagePoppedBefore += block.numL1Messages(totalL1MessagePoppedBefore)
+
+		if len(blockBytes) != 60 {
+			return nil, fmt.Errorf("block encoding is not 60 bytes long %x", len(blockBytes))
+		}
+
+		chunkBytes = append(chunkBytes, blockBytes...)
+
+		// Append rlp-encoded l2Txs
+		for _, txData := range block.Transactions {
+			if txData.Type == types.L1MessageTxType {
+				continue
+			}
+			rlpTxData, err := convertTxDataToRLPEncoding(txData)
+			if err != nil {
+				return nil, err
+			}
+			var txLen [4]byte
+			binary.BigEndian.PutUint32(txLen[:], uint32(len(rlpTxData)))
+			l2TxDataBytes = append(l2TxDataBytes, txLen[:]...)
+			l2TxDataBytes = append(l2TxDataBytes, rlpTxData...)
+		}
+	}
+
+	chunkBytes = append(chunkBytes, l2TxDataBytes...)
+
+	return chunkBytes, nil
+}
+
+// Hash hashes the Chunk into RollupV2 Chunk Hash
+func (c *Chunk) Hash(totalL1MessagePoppedBefore uint64) (common.Hash, error) {
+	chunkBytes, err := c.Encode(totalL1MessagePoppedBefore)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	numBlocks := chunkBytes[0]
+
+	// concatenate block contexts
+	var dataBytes []byte
+	for i := 0; i < int(numBlocks); i++ {
+		// only the first 58 bytes of each BlockContext are needed for the hashing process
+		dataBytes = append(dataBytes, chunkBytes[1+60*i:60*i+59]...)
+	}
+
+	// concatenate l1 and l2 tx hashes
+	for _, block := range c.Blocks {
+		var l1TxHashes []byte
+		var l2TxHashes []byte
+		for _, txData := range block.Transactions {
+			txHash := strings.TrimPrefix(txData.TxHash, "0x")
+			hashBytes, err := hex.DecodeString(txHash)
+			if err != nil {
+				return common.Hash{}, err
+			}
+			if txData.Type == types.L1MessageTxType {
+				l1TxHashes = append(l1TxHashes, hashBytes...)
+			} else {
+				l2TxHashes = append(l2TxHashes, hashBytes...)
+			}
+		}
+		dataBytes = append(dataBytes, l1TxHashes...)
+		dataBytes = append(dataBytes, l2TxHashes...)
+	}
+
+	hash := crypto.Keccak256Hash(dataBytes)
+	return hash, nil
+}
+
+// DecodeChunkBlockRanges decodes the provided chunks into a list of block ranges. Each chunk
+// contains information about multiple blocks, which are decoded and their ranges (from the
+// start block to the end block) are returned.
+func DecodeChunkBlockRanges(chunks [][]byte) ([]*rawdb.ChunkBlockRange, error) {
+	var chunkRanges []*rawdb.ChunkBlockRange
```

Comment:
```suggestion
	var chunkBlockRanges []*rawdb.ChunkBlockRange
```

Reply:
renamed in [65629da](https://github.com/scroll-tech/go-ethereum/pull/515/commits/65629daef071314a5271b91e15672cf64dce1454)